### ISSUE TITLE
Filter entries

### ISSUE THREAD

**Issue by brokkr** [brokkr/poca#29]
Similar to other restrictions on combo.lst we could restrict it further by filtering based on 

- feed info
- filename
- size
- date and time

**Comment by brokkr** [comment]
a) we should create a new object (self.filtered) based off of combo that gathers all the restrictions we're placing on combo (bookmarks, max_no, ?) (or should it only be about filters?)
b) start y using info we already have
c) read up on what elements we can count on: https://pythonhosted.org/feedparser/common-rss-elements.html

**Comment by brokkr** [comment]
For filenames we can use RE and filter, like so:

This is fnlst:

`['http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-05-10-00-3.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-02-14-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-02-13-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-02-12-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-02-10-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-01-14-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-01-13-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-01-12-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-09-01-10-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-31-14-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-31-13-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-31-12-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-31-10-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-30-14-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-30-13-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-30-12-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-30-10-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-29-14-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-29-13-00.mp3?highestBitrate=True&podcastDownload=True',
 'http://www.dr.dk/mu/MediaRedirector/WithFileExtension/radioavisen-2016-08-29-12-00.mp3?highestBitrate=True&podcastDownload=True']
`

We want to find the 12 o'clock entries so we cook up a lambda that evaluates if a regex that looks for '12-00' checks out:

`mlambda = lambda x: bool(re.search('\d{4}-\d{2}-\d{2}-12-00', x))`

And then we apply that lambda as a filter on the whole list of filenames:

`filtered_list = list(filter(mlambda, fnlst))`

The RE in question should obviously come from the user's subscription configuration but seeing that it's just a string, that should not be a problem.

**Comment by brokkr** [comment]
For file sizes we use the same filter logic only the lambda is simpler:

` intlist = [ 0, 12, 5, 7, 123, 88, 22, 65, 81, 11, 2, 49]`
`nlambda = lambda x: x > 33`
` list(filter(nlambda, intlist))`

**Comment by brokkr** [comment]
The question then is how do we structure this?
```
<filters>
    <regex>
        <filename>...</filename>
        <title>...</title>
    </regex>
    <size>
        <less_than>...</less_than>
        <greater_than>...</greater_than>
    </size>
    <hour>
        <less_than>...</less_than>
        <greater_than>...</greater_than>
    </hour>
    <weekday>...</weekday>
</filters>
```
seems a bit much if all we (I) really want to do is filter on filenames...

This can be really powerful but also really really overkill. I guess it depends on how much more complex the code will be if want to allow the user to filter on anything. We could of course just off with a simple:

`<filter>...</filter>`

and then that's just regex filtering on filename?

Q: Can you come up with a case where filename filtering does not suffice?
A: Yes, not all news shows differentiate their daily offerings. If you only want the morning's Newshour you have to use the timestamp rather than the filename.

**Comment by brokkr** [comment]
Here's the shortlist:

```
<filters>
    <filename>...</filename>
    <title>...</title>
    <hour>...</hour>
    <weekday>...</weekday>
</filters>
```

filename and title are regexes, hour and weekday is hour/weekday extracted from pubDate. We'll need re for the first two, but for the latter two we shouldn't need anything since the time objects feedparser delivers have methods to hand us the weekday and hour as integers. That we can simply compare to what the user has entered as criteria.

I think this is a reasonable compromise between power and manageability. It only requires two functions (filename and title checking work similarly, same for hour/wday) and one module. It encapsulates most use cases that I can imagine - I only want a specific type of episode, I only want the ones on fridays or mornings, etc. 

Of course, we don't get the mirror versions, then: Exclude friday episodes, morning episodes, the guest host episodes, etc. Can't really see a simple way to do that, though...

**Comment by brokkr** [comment]
Add filter: Only episodes newer than...

**Comment by brokkr** [comment]
Filters on 

- filename
- title
- hour
- weekday
- after_date

implemented (commit #192)
